### PR TITLE
Refactoring and new features

### PR DESCRIPTION
**Pull request by Kaemmelot**
Since I had some troubles with the current version, I started to redo some parts and ended up refactoring all.

Includes:
- Secure connections
- Improved UI
- Lazy updates
- Node 12 in Docker
- Dimension (Nether/End) support
- Config section for client.js
- Error fixes (invalid package.json, browser crashes, missing catches,...)

-> Moved to version 2.0.0 due to breaking changes

### CODE REVIEW

**Comment by Kaemmelot** [comment]
Would like to hear your opinion on this @ArchmageInc :)

**Comment by ArchmageInc** [comment]
Well... I don't see any unit test failures. lol. I'll try to pull this down and run it because there are too many changes to read through and see what it's doing.

File: server.js [hunk 1]
```diff
@@ -1,194 +1,207 @@
...
+
+  // send
+  if (config.debug)
+    console.info(`Sending player data to ${connections.length} connections`, playerData);
+  connections.forEach((conn) => {
+    conn.send(playerData);
+  });
+}
+
+const playerListRegex = /There are \d+ of a max \d+ players online: ((?:[^\s,]+,?\s?)*)/
```
Comment (TheDruidsKeeper) [inline server.js]: In 1.16 the response says `a max of ` (your regex is missing the "of"). I don't know how this compares to older versions, but your change here breaks the current version.

File: server.js [hunk 2]
```diff
@@ -1,194 +1,207 @@
...
+const playerPosRegex = /has the following entity data: \[(-?\d+.?\d*)d, (-?\d+.?\d*)d, (-?\d+.?\d*)d\]/
+const playerDimRegex = /has the following entity data: (-?\d+)/
+async function updatePlayerData() {
+  // list players
+  const listResponse = await rcon.send('list');
+  let regexMatch = playerListRegex.exec(listResponse);
+  if (regexMatch === null)
+    throw new Error('Invalid response');
```
Comment (TheDruidsKeeper) [inline server.js]: It would be useful for each of these errors to include the response message that it failed to parse.

File: server.js [hunk 3]
```diff
@@ -1,194 +1,207 @@
...
+  if (config.debug)
+    console.info(`Sending player data to ${connections.length} connections`, playerData);
+  connections.forEach((conn) => {
+    conn.send(playerData);
+  });
+}
+
+const playerListRegex = /There are \d+ of a max \d+ players online: ((?:[^\s,]+,?\s?)*)/
+const playerPosRegex = /has the following entity data: \[(-?\d+.?\d*)d, (-?\d+.?\d*)d, (-?\d+.?\d*)d\]/
+const playerDimRegex = /has the following entity data: (-?\d+)/
```
Comment (TheDruidsKeeper) [inline server.js]: The current version of minecraft returns `entity data: "minecraft:overworld"`  (similar for nether/end), not a number, for the dimension.

**Review by TheDruidsKeeper — CHANGES_REQUESTED** [review]
I tried running a build of this with the current version of minecraft (1.16.1) and found several issues that prevent this from working.

**Comment by TheDruidsKeeper** [comment]
@Kaemmelot in the future I would suggest breaking up your PR's into smaller chunks next time, or at the very least make sure you commit your changes to your branch in smaller functional chunks - that would make code reviews a lot easier.

**Comment by Kaemmelot** [comment]
Thanks for the feedback @TheDruidsKeeper.
> in the future I would suggest breaking up your PR's into smaller chunks next time

Sorry about that, when I started making changes I didn't intend to make a PR and afterwards I felt too lazy to redo all steps.

I fixed a few things and added 1.16 compatibility (even though Overviewer itself is not ready for it). Tested with Spigot 1.13 up to 1.16.1.

**Review by TheDruidsKeeper — APPROVED** [review]

**Comment by TheDruidsKeeper** [comment]
@ArchmageInc Any chance you could grant me the access to this repo so I can fix & complete this PR? I've been using @Kaemmelot's fork this whole time along with some additional changes that I'd like to submit as an additional PR.

**Comment by ArchmageInc** [comment]
> @ArchmageInc Any chance you could grant me the access to this repo so I can fix & complete this PR? I've been using @Kaemmelot's fork this whole time along with some additional changes that I'd like to submit as an additional PR.

Done, though I consider this a dead project. Minecraft's RCON interface is terrible. I have split into two projects:

The Overviewer Plugin
https://github.com/ArchmageInc/Overviewer-Info-Plugin

The Spigot Plugin:
https://github.com/ArchmageInc/PlayerLocations

**Pull request merged**